common/set.h

```diff
@@ -0,0 +1,334 @@
+  auto Lookup(LookupKeyT lookup_key,
+              KeyContextT key_context = KeyContextT()) const -> LookupResult {
+    return ViewT(*this).Lookup(lookup_key, key_context);
+  }
+
+  // Convenience forwarder to the view type.
+  template <typename CallbackT>
+  void ForEach(CallbackT callback) {
+    return ViewT(*this).ForEach(callback);
+  }
+
+  // Convenience forwarder to the view type.
+  auto CountProbedKeys(KeyContextT key_context = KeyContextT()) const
+      -> ssize_t {
+    return ViewT(*this).CountProbedKeys(key_context);
+  }
+
+  // Insert a key into the set. If the key is already present, no insertion is
+  // performed and that present key is available in the result. Otherwise a new
+  // key is inserted and constructed from the argument and available in the
+  // result.
+  template <typename LookupKeyT>
+  auto Insert(LookupKeyT lookup_key, KeyContextT key_context = KeyContextT())
+      -> InsertResult;
+
+  // Insert a key into the set and call the provided callback to allow in-place
+  // construction of the key if not already present. The lookup key is passed
+  // through to the callback so it needn't be captured and can be kept in a
+  // register argument throughout.
+  //
+  // Example:
+  // ```cpp
+  //   m.Insert("widget", [](MyStringViewType lookup_key, void* key_storage) {
+  //     new (key_storage) MyStringType(lookup_key);
+  //   });
+  // ```
+  template <typename LookupKeyT, typename InsertCallbackT>
+  auto Insert(LookupKeyT lookup_key, InsertCallbackT insert_cb,
+              KeyContextT key_context = KeyContextT()) -> InsertResult
+    requires std::invocable<InsertCallbackT, LookupKeyT, void*>;
+
+  // Erase a key from the set.
+  template <typename LookupKeyT>
+  auto Erase(LookupKeyT lookup_key, KeyContextT key_context = KeyContextT())
+      -> bool;
+
+  // Clear all key/value pairs from the set but leave the underlying hashtable
+  // allocated and in place.
+  void Clear();
+
+ protected:
+  using ImplT::ImplT;
+};
+
+// A data structure for a set of keys.
+//
+// This set also supports small size optimization (or "SSO"). The provided
+// `SmallSize` type parameter indicates the size of an embedded buffer for
+// storing sets small enough to fit. The default is zero, which always allocates
+// a heap buffer on construction. When non-zero, must be a multiple of the
+// `MaxGroupSize` which is currently 16. The library will check that the size is
+// valid and provide an error at compile time if not. We don't automatically
+// select the next multiple or otherwise fit the size to the constraints to make
+// it clear in the code how much memory is used by the SSO buffer.
+//
+// This data structure optimizes heavily for small key types that are cheap to
+// move and even copy. Using types with large keys or expensive to copy keys may
+// create surprising performance bottlenecks. A `std::string` key should be fine
+// with largely small strings, but if some or many strings are large heap
```

Comment:
"largely" seems strange to describe something small
```suggestion
// with generally small strings, but if some or many strings are large heap
```

Reply:
Done.